### PR TITLE
OnboardingDialog

### PR DESCRIPTION
**Pull request by glenngijsberts**
# Description

@jelleoverbeek presented a new dialog design to use for forced onboarding. 

## Changes

- [x] Add new component
- [x] Add story

## How to test

- Checkout this branch
- `$ yarn dev`
- Checkout the story 'OnboardingDialog'.

## Screenshots

Example usage:

Screenshot 2020-07-15 at 10 58 06

### CODE REVIEW

File: src/components/ContentTransition/index.js [hunk 1]
```diff
@@ -90,7 +90,11 @@ export function ContentTransition({ children, onChange, activeView }) {
...
 }
 
+ContentTransition.defaultProps = {
+  onChange: () => null,
+}
+
 ContentTransition.propTypes = {
   activeView: PropTypes.string.isRequired,
-  onChange: PropTypes.func.isRequired,
+  onChange: PropTypes.func,
```
Comment (RobVermeer) [inline src/components/ContentTransition/index.js:99]: This change should be its own commit.

File: src/components/OnboardingDialog/index.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+const OnboardingDialogBody = ({ children }) => (
+  <BodyWrapper>{children}</BodyWrapper>
+)
+
+OnboardingDialogBody.propTypes = {
+  children: PropTypes.oneOfType([
+    PropTypes.arrayOf(PropTypes.node),
+    PropTypes.node,
+  ]).isRequired,
+}
```
Comment (RobVermeer) [inline src/components/OnboardingDialog/index.js:32]: I don't think we need propType validation for children 🤔

File: src/components/OnboardingDialog/index.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+const DialogWrapper = styled.div`
+  position: absolute;
+  top: ${props => space[props.positionTop]};
```
Comment (RobVermeer) [inline src/components/OnboardingDialog/index.js]: Position top is also dependant on screen-size, so maybe remove the prop, and just do fixed for mobile/desktop 🤔

File: src/components/OnboardingDialog/index.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  positionTop: 56,
+}
+
+OnboardingDialog.propTypes = {
+  positionTop: PropTypes.number,
+  show: PropTypes.bool.isRequired,
+  children: PropTypes.oneOfType([
+    PropTypes.arrayOf(PropTypes.node),
+    PropTypes.node,
+  ]).isRequired,
```
Comment (RobVermeer) [inline src/components/OnboardingDialog/index.js:61]: Same here, not needed

File: src/components/OnboardingDialog/index.js [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  position: relative;
+`
+
+const Backdrop = styled.div`
+  position: absolute;
+  top: 0;
+  left: 0;
+  right: 0;
+  bottom: 0;
+  background-color: rgba(255, 255, 255, 0.8);
```
Comment (RobVermeer) [inline src/components/OnboardingDialog/index.js]: Dow we already have something similar, opacity-wise? 🤔 Or maybe this should be within the theme.

File: src/components/OnboardingDialog/index.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+  )
+}
+
+OnboardingWrapper.propTypes = {
+  active: PropTypes.bool.isRequired,
+  state: PropTypes.string.isRequired,
+  children: PropTypes.oneOfType([
+    PropTypes.arrayOf(PropTypes.node),
+    PropTypes.node,
+  ]).isRequired,
```
Comment (RobVermeer) [inline src/components/OnboardingDialog/index.js:95]: Same

File: src/components/OnboardingDialog/index.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    <Container>
+      {children}
+      {active && <Backdrop state={state} />}
+    </Container>
+  )
+}
+
+OnboardingWrapper.propTypes = {
+  active: PropTypes.bool.isRequired,
+  state: PropTypes.string.isRequired,
```
Comment (RobVermeer) [inline src/components/OnboardingDialog/index.js]: This should be a `oneOf(['ENTERED', 'xxx etc'])`

File: src/components/OnboardingDialog/index.js [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+Onboarding.defaultProps = {
+  defaultOn: true,
+}
+
+Onboarding.propTypes = {
+  defaultOn: PropTypes.bool.isRequired,
+  children: PropTypes.oneOfType([
+    PropTypes.arrayOf(PropTypes.node),
+    PropTypes.node,
+  ]).isRequired,
```
Comment (RobVermeer) [inline src/components/OnboardingDialog/index.js:115]: Same

File: src/components/OnboardingDialog/index.js [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+function useOnboarding(config = {}) {
+  const { defaultOn = true, defaultTransitionState = 'exited' } = config
```
Comment (RobVermeer) [inline src/components/OnboardingDialog/index.js]: `exited` should be a const, see also other implementations. Maybe we can import them from `useTransition`

Comment (RobVermeer) [reply src/components/OnboardingDialog/index.js]: Btw we then could use those consts in the propType validation as stated above as well.

File: src/components/OnboardingDialog/index.stories.js [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+const Actions = styled.div`
+  display: grid;
+  margin-top: ${space[16]};
+  grid-template-columns: 200px 200px;
+  grid-gap: ${space[8]};
+`
+
+const StorybookDialog = () => {
+  const { show, hide, getOnboardingProps, state, active } = useOnboarding({
+    defaultTransitionState: 'entered',
```
Comment (RobVermeer) [inline src/components/OnboardingDialog/index.stories.js:35]: Is the prop `defaultTransitionState` really needed, could you explain what it actually does?

File: src/components/OnboardingDialog/index.js (hunk 2 above)
Comment (glenngijsberts) [reply src/components/OnboardingDialog/index.js:32]: Why not?

Comment (glenngijsberts) [reply src/components/OnboardingDialog/index.js:32]: Since it _is_ a prop, I think it does no harm to add it. Also, it will remind you if you forget to include it. What do you think, @Roohn?

File: src/components/OnboardingDialog/index.js (hunk 3 above)
Comment (glenngijsberts) [reply src/components/OnboardingDialog/index.js]: Isn't dependant on the child content? I've added it to be able to make adjustments to fit it better with underlaying content.

File: src/components/OnboardingDialog/index.js (hunk 5 above)
Comment (glenngijsberts) [reply src/components/OnboardingDialog/index.js]: I'll check it out.

File: src/components/OnboardingDialog/index.js (hunk 2 above)
Comment (Roohn) [reply src/components/OnboardingDialog/index.js:32]: Well that is a good one, I never do it because the PropTypes are always the same.
But I see no harm in doing this. Maybe it will be good to do this in every component

**Comment by Roohn** [comment]
@glenngijsberts. Recently we added [@reach/dialog](https://reach.tech/dialog). I think that dialog would be good to use for this. Since you only need to add some styling to it and the accessibility would be better.

**Comment by Roohn** [comment]
Also we have this issue https://github.com/TicketSwap/solar/issues/536 
And this would make the issue even bigger 😅

File: src/components/OnboardingDialog/index.js (hunk 2 above)
Comment (glenngijsberts) [reply src/components/OnboardingDialog/index.js:32]: Yes let's do that!

File: src/components/OnboardingDialog/index.stories.js (hunk 10 above)
Comment (glenngijsberts) [reply src/components/OnboardingDialog/index.stories.js:35]: In the current implementation the default state is hard coded with 'exited'. This works fine with our default dialogs, but with this specific dialog it will clearly show that the backdrop will fade in a bit after the page load which looks awful.

File: src/components/OnboardingDialog/index.js (hunk 5 above)
Comment (glenngijsberts) [reply src/components/OnboardingDialog/index.js]: I'll add it to the theme. I'll call it `novaAlpha`.

File: src/components/OnboardingDialog/index.js (hunk 3 above)
Comment (RobVermeer) [reply src/components/OnboardingDialog/index.js]: Then we should make it possible to make a styled DialogWrapper. You can do this by spreading the rest props on the wrapper.

Comment (glenngijsberts) [reply src/components/OnboardingDialog/index.js]: Resolved by passing props to DialogWrapper so you can use a styled component instead of you want to alter it.

**Comment by glenngijsberts** [comment]
@Roohn I've looked into it and it seemed that it wasn't really fitting the dialog we created. Rob and I decided to implement the same attributes that Reach is using for the dialog (wrapper). The change is included in my latest commit.

File: src/theme.js [hunk 11]
```diff
@@ -12,6 +12,9 @@ export const color = {
   spaceLighterAlpha: 'rgba(0,19,25,0.2)',
   spaceLightestAlpha: 'rgba(0,19,25,0.1)',
 
+  nova: 'rgb(255, 255, 255)',
+  novaAlpha: 'rgba(255,255,255,0.8)',
```
Comment (RobVermeer) [inline src/theme.js]: These are all the colors that are defined in the design system?

Comment (glenngijsberts) [reply src/theme.js]: I'll ask Jelle.

File: src/hooks/useTransition.js [hunk 12]
```diff
@@ -1,13 +1,14 @@
...
 import { usePrevious } from './usePrevious'
 
-const EXITED = 'exited'
-const ENTERING = 'entering'
-const ENTERED = 'entered'
-const EXITING = 'exiting'
+export const EXITED = 'exited'
+export const ENTERING = 'entering'
+export const ENTERED = 'entered'
+export const EXITING = 'exiting'
```
Comment (RobVermeer) [inline src/hooks/useTransition.js:7]: Check the rest of the codebase, we can use these constants there as well (f.e. dialog, content transition)

File: src/components/OnboardingDialog/index.js [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+  return (
+    <DialogWrapper aria-modal="true" role="dialog" tabindex="-1" {...props}>
+      {children}
+    </DialogWrapper>
+  )
+}
+
+OnboardingDialog.defaultProps = {
+  positionTop: 56,
+}
```
Comment (RobVermeer) [inline src/components/OnboardingDialog/index.js]: Can be removed

File: src/components/OnboardingDialog/index.js [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+
+OnboardingDialog.propTypes = {
+  positionTop: PropTypes.number,
```
Comment (RobVermeer) [inline src/components/OnboardingDialog/index.js]: Can be removed

File: src/components/OnboardingDialog/index.js [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+import { EXITED, ENTERING, ENTERED, EXITING } from '../../hooks/useTransition'
+
+const BodyWrapper = styled.div`
+  background-color: white;
+  border-radius: ${radius.lg};
+  box-shadow: ${shadow.strong};
+  padding: ${space[32]};
+`
+
+const OnboardingDialogBody = ({ children }) => (
```
Comment (RobVermeer) [inline src/components/OnboardingDialog/index.js]: I would add `...props` here as well, since this is exported and thus needs styled components support.

File: src/components/OnboardingDialog/index.js [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+  left: 0;
+  right: 0;
+  bottom: 0;
+  background-color: ${color.novaAlpha};
+  opacity: ${props =>
+    props.state === ENTERING || props.state === ENTERED ? 1 : 0};
+  transition: opacity ${transition};
+`
+
+const OnboardingWrapper = ({ children, active, state }) => {
```
Comment (RobVermeer) [inline src/components/OnboardingDialog/index.js]: I would add `...props` here as well, since this is exported and thus needs styled components support.

File: src/components/OnboardingDialog/index.js [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+  duration,
+  transition,
+  color,
+  device,
+} from '../../theme'
+import { useTransition } from '../../hooks'
+import { EXITED, ENTERING, ENTERED, EXITING } from '../../hooks/useTransition'
+
+const BodyWrapper = styled.div`
+  background-color: white;
```
Comment (RobVermeer) [inline src/components/OnboardingDialog/index.js:17]: `color.nova` (maybe good to do this for everything is a separate commit)

Comment (glenngijsberts) [reply src/components/OnboardingDialog/index.js:17]: I would rather do that in another PR.

File: src/hooks/useTransition.js (hunk 12 above)
Comment (glenngijsberts) [reply src/hooks/useTransition.js:7]: I'll do this in a separate PR, so I can test everything and make sure it still works 😅

**Review by RobVermeer — APPROVED** [review]

**Review by Roohn — APPROVED** [review]
